### PR TITLE
[DM] [NINOA-44] Added Name model

### PR DESCRIPTION
**Pull request by TheDeadlyPianist**

### CODE REVIEW

File: test/v1/models/NameModelSpec.scala [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        val expectedMessage = "Unable to validate forename. Ensure it matches the regex."
+
+        val exceptionResultInvalidCharacters = intercept[JsResultException] {
+          dodgeyJsonCharacters.as[NameModel]
+        }
+        val exceptionResultInvalidLength = intercept[JsResultException] {
+          dodgeyJsonLength.as[NameModel]
+        }
+
+        exceptionResultInvalidCharacters.getMessage should include(expectedMessage)
```
Comment (TomDudley92) [inline test/v1/models/NameModelSpec.scala]: Can you put these into separate tests please? This also goes for the secondForename and surname field tests as well please.

File: app/v1/models/NameModel.scala [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    titleInput.filter(
+      JsonValidationError("Title failed validation. Must be one of: \'NOT KNOWN, MR, MRS, MISS, MS, DR, REV\'"))(
+      titleInputString => validationCheck(titleInputString)
+    )
+  }
+
+  private val nameRegex = "^(?=.{1,35}$)([A-Z]([-'.&\\\\/ ]{0,1}[A-Za-z]+)*[A-Za-z]?)$"
+  private def jsError: String => JsonValidationError = fieldName => JsonValidationError(s"Unable to validate $fieldName. Ensure it matches the regex.")
+
+  private def checkAgainstNameRegex(nameString: String): Boolean = {
```
Comment (TomDudley92) [inline app/v1/models/NameModel.scala]: Does this need to be it's own def? It's only used once

File: test/v1/models/NameModelSpec.scala [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+          "startDate" -> "10-10-2020"
+        )
+
+        val exceptionResult = intercept[JsResultException] {
+          dodgeyJson.as[NameModel]
+        }
+
+        exceptionResult.getMessage should include("Title failed validation. Must be one of: 'NOT KNOWN, MR, MRS, MISS, MS, DR, REV'")
+      }
+      "the forename field fails validation" in {
```
Comment (TomDudley92) [inline test/v1/models/NameModelSpec.scala]: This test could just test <private def validateName>. If you did the same tests on <validateName> you wouldn't need to test each individual field as you already have the confidence that the method will work.

File: test/v1/models/NameModelSpec.scala [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  "NameModelSpec" should {
+    "correctly parse to Json" in {
+      Json.toJson(validModel) shouldBe validJsonWrite
+    }
+    "correctly parse from Json" in {
+      validJsonRead.as[NameModel] shouldBe validModel
+    }
+
+    "throw an error when parsing from json" when {
+      "the Title field fails validation" in {
```
Comment (TomDudley92) [inline test/v1/models/NameModelSpec.scala]: Can you test the validateTitle method for instead of reading it as a model please?

File: test/v1/models/NameModelSpec.scala [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1.models
+
+import play.api.libs.json.{JsObject, JsResultException, Json}
+import support.UnitSpec
+
+class NameModelSpec extends UnitSpec {
```
Comment (TomDudley92) [inline test/v1/models/NameModelSpec.scala]: For this spec can you add a test for creating a minimum and maximum viable model please?

File: app/v1/models/NameModel.scala [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    "MISS",
+    "MS",
+    "DR",
+    "REV"
+  )
+
+  private[models] def validateTitle: Option[String] => Boolean = {
+    case Some(inputString) =>
+      val passedValidation = validTitles.contains(inputString)
+      if (!passedValidation) Logger.warn(s"[NameModel][]validateTitle] The following title failed validation: $inputString")
```
Comment (TomDudley92) [inline app/v1/models/NameModel.scala]: There is an extra 


```suggestion
      if (!passedValidation) Logger.warn(s"[NameModel][validateTitle] The following title failed validation: $inputString")
```
